File: Lecture.3/gcd.py

# use py to interactively run this file
# use exec(open("gcd.py").read()) to dynamically load defintions in this file or directly run 'py -i .\gcd.py'
# run by  executing: 
#  >>> gcd_it(10,20)
#  >>> gcd_rec(10,20)
#  >>> gcd_fun(10,20)


# iterative ugly
def gcd_it_ugly(a,b):
    while  a != 0 and b != 0:
        if a > b:
            a = a - b
        else:
            b = b - a
    
    if a == 0:
        return b
    else:
        return a
    


# iterative accelerated algorithm
def gcd_it(a,b):
    while b:
        a,b = b, a%b # pattern matching expression
    return a

# recursive approach
def gcd_rec(a,b):
    return a if not b else gcd_rec(b, a%b)

# functional approach 
gcd_fun = lambda a,b : a if not b else gcd_fun(b, a%b)

File: Lecture.3/gcd.pl
% use swipl to interactively run this file
% use [gcd]. to load the definitions in this file
% run using 
%   - gcd(100,230,X).
%   - gcd_euc(100,230,X).
%   - gcdl([100,200,300], X).
% use halt. to quit

% slow approach using Euclides' original algorithm
gcd_euc(0, X, X):- X > 0, !.
gcd_euc(X, Y, Z):- X >= Y, X1 is X-Y, gcd_euc(X1,Y,Z).
gcd_euc(X, Y, Z):- X < Y, X1 is Y-X, gcd_euc(X1,X,Z).


% accelerated algorithm using mod operator
gcd(X, Y, X):- Y = 0, !.
gcd(X, Y, Z):- X1 is (X mod Y), gcd(Y, X1, Z).


% gcd of a list
gcdl([X1,X2|T], Z) :- gcd(X1,X2,X), gcdl([X|T], Z).
gcdl([X1,X2], Z):- gcd(X1,X2,Z).
